src/libstd/io/impls.rs

```diff
@@ -0,0 +1,73 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use core::prelude::*;
+
+use cmp;
+use io::{self, SeekFrom, Read, Write, Seek, BufRead};
+use ptr;
+use slice;
+use vec::Vec;
+
+// =============================================================================
+// Forwarding implementations
+
+impl<'a, R: Read> Read for &'a mut R {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> { (**self).read(buf) }
+}
+impl<'a, W: Write> Write for &'a mut W {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> { (**self).write(buf) }
+    fn flush(&mut self) -> io::Result<()> { (**self).flush() }
+}
+impl<'a, S: Seek> Seek for &'a mut S {
+    fn seek(&mut self, pos: SeekFrom) -> io::Result<u64> { (**self).seek(pos) }
+}
+impl<'a, B: BufRead> BufRead for &'a mut B {
+    fn fill_buf(&mut self) -> io::Result<&[u8]> { (**self).fill_buf() }
+    fn consume(&mut self, amt: usize) { (**self).consume(amt) }
+}
+
+// =============================================================================
+// In-memory buffer implementations
+
+impl<'a> Read for &'a [u8] {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let amt = cmp::min(buf.len(), self.len());
+        let (a, b) = self.split_at(amt);
+        slice::bytes::copy_memory(buf, a);
+        *self = b;
+        Ok(amt)
+    }
+}
+
+impl<'a> BufRead for &'a [u8] {
+    fn fill_buf(&mut self) -> io::Result<&[u8]> { Ok(*self) }
+    fn consume(&mut self, amt: usize) { *self = &self[amt..]; }
+}
+
+impl<'a> Write for &'a mut [u8] {
+    fn write(&mut self, data: &[u8]) -> io::Result<usize> {
+        let amt = cmp::min(data.len(), self.len());
+        slice::bytes::copy_memory(*self, &data[..amt]);
+        unsafe {
+            let other = ptr::read(self);
+            *self = &mut other[amt..];
+        }
```

Comment:
This works ([see in playpen](https://play.rust-lang.org/?code=use%20std%3A%3Amem%3B%0A%0Afn%20cut%3CT%3E%28xs%3A%20%26mut%20%26mut%20%5BT%5D%2C%20amt%3A%20usize%29%20%7B%0A%20%20%20%20let%20original%20%3D%20mem%3A%3Areplace%28xs%2C%20%26mut%20%5B%5D%29%3B%0A%20%20%20%20mem%3A%3Areplace%28xs%2C%20%26mut%20original%5Bamt..%5D%29%3B%0A%7D%0A%0Afn%20main%28%29%20%7B%0A%20%20%20%20let%20mut%20xs%3A%20%26mut%20%5B_%5D%20%3D%20%26mut%20%5B1%2C%202%2C%203%5D%3B%0A%20%20%20%20cut%28%26mut%20xs%2C%202%29%3B%0A%20%20%20%20println%21%28%22%7B%3A%3F%7D%22%2C%20xs%29%3B%0A%7D)):
```rust
fn cut<T>(xs: &mut &mut [T], amt: usize) {
    let original = mem::replace(xs, &mut []);
    mem::replace(xs, &mut original[amt..]);
}
```